libcxx/test/std/containers/container.adaptors/flat.map/flat.map.cons/move_assign.pass.cpp

```diff
@@ -0,0 +1,110 @@
+
+#include <algorithm>
+#include <deque>
+#include <flat_map>
+#include <functional>
+#include <memory_resource>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "test_macros.h"
+#include "MoveOnly.h"
+#include "../../../test_compare.h"
+#include "test_allocator.h"
+#include "min_allocator.h"
+
+int main(int, char**) {
+  {
+    using C  = test_less<int>;
+    using A1 = test_allocator<int>;
+    using A2 = test_allocator<char>;
+    using M  = std::flat_map<int, char, C, std::vector<int, A1>, std::vector<char, A2>>;
+    M mo     = M({{1, 1}, {2, 3}, {3, 2}}, C(5), A1(7));
+    M m      = M({}, C(3), A1(7));
+    m        = std::move(mo);
+    assert((m == M{{1, 1}, {2, 3}, {3, 2}}));
+    assert(m.key_comp() == C(5));
+    auto [ks, vs] = std::move(m).extract();
+    assert(ks.get_allocator() == A1(7));
+    assert(vs.get_allocator() == A2(7));
+    assert(mo.empty());
+  }
+  {
+    using C  = test_less<int>;
+    using A1 = other_allocator<int>;
+    using A2 = other_allocator<char>;
+    using M  = std::flat_map<int, char, C, std::deque<int, A1>, std::deque<char, A2>>;
+    M mo     = M({{4, 5}, {5, 4}}, C(5), A1(7));
+    M m      = M({{1, 1}, {2, 2}, {3, 3}, {4, 4}}, C(3), A1(7));
+    m        = std::move(mo);
+    assert((m == M{{4, 5}, {5, 4}}));
+    assert(m.key_comp() == C(5));
+    auto [ks, vs] = std::move(m).extract();
+    assert(ks.get_allocator() == A1(7));
+    assert(vs.get_allocator() == A2(7));
+    assert(mo.empty());
+  }
+  {
+    using A = min_allocator<int>;
+    using M = std::flat_map<int, int, std::greater<int>, std::vector<int, A>, std::vector<int, A>>;
+    M mo    = M({{5, 1}, {4, 2}, {3, 3}}, A());
+    M m     = M({{4, 4}, {3, 3}, {2, 2}, {1, 1}}, A());
+    m       = std::move(mo);
+    assert((m == M{{5, 1}, {4, 2}, {3, 3}}));
+    auto [ks, vs] = std::move(m).extract();
+    assert(ks.get_allocator() == A());
+    assert(vs.get_allocator() == A());
+    assert(mo.empty());
+  }
+  {
+    // A moved-from flat_map maintains its class invariant in the presence of moved-from elements.
+    using M =
+        std::flat_map<std::pmr::string, int, std::less<>, std::pmr::vector<std::pmr::string>, std::pmr::vector<int>>;
+    std::pmr::monotonic_buffer_resource mr1;
+    std::pmr::monotonic_buffer_resource mr2;
+    M mo = M({{"short", 1},
+              {"very long string that definitely won't fit in the SSO buffer and therefore becomes empty on move", 2}},
+             &mr1);
+    M m  = M({{"don't care", 3}}, &mr2);
+    m    = std::move(mo);
+    assert(m.size() == 2);
+    assert(std::is_sorted(m.begin(), m.end(), m.value_comp()));
+    assert(m.begin()->first.get_allocator().resource() == &mr2);
+
+    assert(std::is_sorted(mo.begin(), mo.end(), mo.value_comp()));
+    mo.insert({"foo", 1});
+    assert(mo.begin()->first.get_allocator().resource() == &mr1);
+  }
+  {
+    // A moved-from flat_map maintains its class invariant in the presence of moved-from comparators.
```

Comment:
I don't think we maintain the invariant since we move the comparator out of the moved-from `flat_map`. Also applies to the move constructor. That makes me wonder how this test is passing.

Reply:
I think the intent of this test was to test that we maintain the invariant when we have a moved-from comparator. But that test is incorrectly checking that, because the comparator fits into the small buffer of the `std::function`, leading to it being trivially movable.

If we fixed this test to check what it intends to check, I would expect it to fail because we don't maintain the invariant (can you check that by using a very large comparator instead?). Something like

```
struct Large : std::less<int> {
  char foo[99999];
};
```

Assuming the explanation above is correct, this should crash cause we don't maintain the invariant.

That being said, instead of trying to fix that test and maintain the invariant, I think we should move the comparator and not attempt to maintain that invariant. A moved-from object can be destroyed or assigned to in the general case, and I'm not aware of anything in the spec that would require that we do more than that for `flat_map`.

Reply:
I can't find something that specifies to what extent the implementation should care about move-from objects. But here  http://eel.is/c++draft/flat.map#overview-6  it says

> If any member function in [[flat.map.defn]](http://eel.is/c++draft/flat.map#defn) exits via an exception the invariants are restored[.](http://eel.is/c++draft/flat.map#overview-6.sentence-1)
[Note [2](http://eel.is/c++draft/flat.map#overview-note-2): This can result in the flat_map being emptied[.](http://eel.is/c++draft/flat.map#overview-6.sentence-2) — end note]

Move constructor/assignment are member functions so I wonder if we should still need to keep its invariant